jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -3640,32 +3717,56 @@ private static boolean counterPredicate(int endValue, int counter) {
 	public static MethodHandle iteratedLoop(MethodHandle iteratorHandle, MethodHandle initHandle, MethodHandle bodyHandle) throws NullPointerException, IllegalArgumentException {
 		validateArgumentsOfIteratedLoop(iteratorHandle, initHandle, bodyHandle);
 		
-		/* Method handles to Iterator.hasNext(), Iterator.next() and Iterable.iterator() are required
-		 * to construct a clause array in the the generic loop.
-		 */
-		MethodHandle iteratorNextElement = null;
-		MethodHandle iteratorHasNextElement = null;
-		try {
-			iteratorNextElement = Lookup.internalPrivilegedLookup.findVirtual(Iterator.class, "next", MethodType.methodType(Object.class)); //$NON-NLS-1$
-			iteratorHasNextElement = Lookup.internalPrivilegedLookup.findVirtual(Iterator.class, "hasNext", MethodType.methodType(boolean.class)); //$NON-NLS-1$
-		} catch (IllegalAccessException | NoSuchMethodException e) {
-			throw new InternalError("The next/hasNext method doesn't exist or it fails in the access checking", e); //$NON-NLS-1$
-		}
+		MethodType bodyType = bodyHandle.type;
+		Class<?> bodyReturnType = bodyType.returnType;
 		
 		/* The init handle for iterator is set to the method handle to Iterable.iterator()
-		 * by default if it is null.
+		 * by default if it is null. The default iterator handle parameters are adjusted 
+		 * to accept the leading A type.
 		 */
+		Class<?> iteratorType;
 		MethodHandle initIterator = iteratorHandle;
 		if (null == initIterator) {
+			iteratorType = Iterator.class;
+			
+			int bodyParamTypesLength = bodyHandle.type.arguments.length;
+			int bodyIterationVarLength = 2;
+			if (void.class == bodyReturnType) {
+				bodyIterationVarLength = 1;
+			}
+			
+			/* If body has an A type, the default iterator handle is adjusted to accept the leading A type. 
+			 * Validation has already confirmed that the leading A type is an Iterable or subtype thereof.
+			 */
+			Class<?> iterableType;
+			if (bodyParamTypesLength > bodyIterationVarLength) {
+				iterableType = bodyHandle.type.arguments[bodyIterationVarLength];
+			} else {
+				iterableType = Iterable.class;
+			}
+			
 			try {
-				initIterator = Lookup.internalPrivilegedLookup.findVirtual(Iterable.class, "iterator", MethodType.methodType(Iterator.class)); //$NON-NLS-1$
+				initIterator = Lookup.internalPrivilegedLookup.findVirtual(iterableType, "iterator", MethodType.methodType(iteratorType)); //$NON-NLS-1$
```

Comment:
The API spec says that:
If iterator is null it defaults to a method handle which behaves like Iterable.iterator(). 
In that case, the internal parameter list (V T A...) must have at least one A type, 
and the default iterator handle parameter is adjusted to accept the leading A type,
as if by the asType conversion method. The leading A type must be Iterable or a subtype thereof. 

So, the default iterator handle must contain at least one A type (Iterable or a subtype, which needs to call asType() to convert).    It seems your iterator handle has no parameter here.
Please check the code here against the constraint above.

Reply:
This is done using the variable `iterableType` which to me seemed like a more elegant solution. I've verified that by calling Lookup on iterable or a subtype of it the first parameter of initIterator will be iterableType which is assigned to the leading A type of bodyHandle if it exists. 

The else case to assign iterableType comes from the rule:
> As a special case, if the body contributes only V and T types, with no additional A types, then the internal parameter list is extended by the argument types A... of the iterator handle; if it is null the single type Iterable is added and constitutes the A... list.